types/wlr_input_method_v2.c

```diff
@@ -101,10 +122,196 @@ static void im_get_input_popup_surface(struct wl_client *client,
+		MAP_SHARED, keymap_fd, 0);
+	if (ptr == MAP_FAILED) {
+		wlr_log(WLR_ERROR, "failed to mmap() %zu bytes",
+			keyboard->keymap_size);
+		close(keymap_fd);
+		return false;
+	}
+
+	memcpy(ptr, keyboard->keymap_string, keyboard->keymap_size);
+	munmap(ptr, keyboard->keymap_size);
+
+	zwp_input_method_keyboard_grab_v2_send_keymap(keyboard_grab->resource,
+		WL_KEYBOARD_KEYMAP_FORMAT_XKB_V1, keymap_fd,
+		keyboard->keymap_size);
+
+	close(keymap_fd);
+	return true;
+}
+
+static void keyboard_grab_send_repeat_info(
+		struct wlr_input_method_keyboard_grab_v2 *keyboard_grab,
+		struct wlr_keyboard *keyboard) {
+	zwp_input_method_keyboard_grab_v2_send_repeat_info(
+		keyboard_grab->resource, keyboard->repeat_info.rate,
+		keyboard->repeat_info.delay);
+}
+
+static void handle_keyboard_keymap(struct wl_listener *listener, void *data) {
+	struct wlr_input_method_keyboard_grab_v2 *keyboard_grab =
+		wl_container_of(listener, keyboard_grab, keymap_listener);
+	keyboard_grab_send_keymap(keyboard_grab, data);
+}
+
+static void handle_keyboard_repeat_info(struct wl_listener *listener,
+		void *data) {
+	struct wlr_input_method_keyboard_grab_v2 *keyboard_grab =
+		wl_container_of(listener, keyboard_grab, keymap_listener);
+	keyboard_grab_send_repeat_info(keyboard_grab, data);
+}
+
+static void handle_keyboard_destroy(struct wl_listener *listener,
+		void *data) {
+	struct wlr_input_method_keyboard_grab_v2 *keyboard_grab =
+		wl_container_of(listener, keyboard_grab, keymap_listener);
+	wlr_input_method_keyboard_grab_v2_set_keyboard(keyboard_grab, NULL);
+}
+
+void wlr_input_method_keyboard_grab_v2_set_keyboard(
+		struct wlr_input_method_keyboard_grab_v2 *keyboard_grab,
+		struct wlr_keyboard *keyboard) {
+	if (keyboard == keyboard_grab->keyboard) {
+		return;
+	}
+
+	if (keyboard_grab->keyboard) {
+		wl_list_remove(&keyboard_grab->keymap_listener.link);
+		wl_list_remove(&keyboard_grab->repeat_info_listener.link);
+		wl_list_remove(&keyboard_grab->destroy_listener.link);
+	}
+
+	if (keyboard) {
+		if (keyboard_grab->keyboard == NULL ||
+				strcmp(keyboard_grab->keyboard->keymap_string,
+				keyboard->keymap_string) != 0) {
+			// send keymap only if it is changed, or if input method is not
+			// aware that it did not change and blindly send it back with
+			// virtual keyboard, it may cause an infinite recursion.
+			if (!keyboard_grab_send_keymap(keyboard_grab, keyboard)) {
+				wlr_log(WLR_ERROR, "Failed to send keymap for input-method keyboard grab");
+				return;
+			}
+		}
+		keyboard_grab_send_repeat_info(keyboard_grab, keyboard);
+		keyboard_grab->keymap_listener.notify = handle_keyboard_keymap;
+		wl_signal_add(&keyboard->events.keymap,
+			&keyboard_grab->keymap_listener);
+		keyboard_grab->repeat_info_listener.notify =
+			handle_keyboard_repeat_info;
+		wl_signal_add(&keyboard->events.repeat_info,
+			&keyboard_grab->repeat_info_listener);
+		keyboard_grab->destroy_listener.notify =
```

Comment:
`wl_signal_add` is not called for this listener.

Reply:
Fixed.